### PR TITLE
Make extjob python only and made `STDIN null` behave the same as `STDOUT null`

### LINKED ISSUE #4337: ExtJob - move C to Python
Blocked by #4354

### PR DESCRIPTION
**Pull request by eivindjahren**
**Issue**
Resolves #4337 

Note, for consistency. This changes the `STDIN null` to behave the same as `STDOUT null`, that is, it disables those files.

## Pre review checklist

- [x] Added appropriate release note label
- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 1]
```diff
@@ -1,263 +1,211 @@
...
+                f"ExtJob {name} has executable set to directory {resolved}"
+            )
 
-        if self.get_stdout_file() != other.get_stdout_file():
-            return False
+        return resolved
 
-        return True
+    def copy(self):
+        return replace(self)
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:79]: Why use replace if none of the dataclass arguments are replaced ?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 2]
```diff
@@ -1,263 +1,211 @@
...
-from ecl.util.util import StringList
-
-from ert._c_wrappers import ResPrototype
-from ert._c_wrappers.config import ContentTypeEnum
+from ert._c_wrappers.config import ConfigParser, ConfigValidationError, ContentTypeEnum
 from ert._c_wrappers.util import SubstitutionList
+from ert._clib.job_kw import type_from_kw
+
+
+@dataclass
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:13]: I am not really sure these types of objects should be forced to be dataclasses as far as I can tell a dataclass should be only a data object that has very little logic (hopefully none) around the data fields as it is implemented now it does not really fit that definition.

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 3]
```diff
@@ -1,263 +1,211 @@
...
+        "EXECUTABLE",
+        "PORTABLE_EXE",
+        "TARGET_FILE",
+        "ERROR_FILE",
+        "START_FILE",
+    ]
+
+    @classmethod
+    def _parse_config_file(cls, config_file: str):
+        parser = ConfigParser()
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:96]: Is it possible to have this `parser = ConfigParser()` created only once or do we need to have a new one each time we call `ExtJob.from_config_file(...)`?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 4]
```diff
@@ -1,263 +1,211 @@
...
+
+        for key in cls._str_keywords:
+            if config_content.hasKey(key):
+                if key in ("STDIN", "STDOUT", "STDERR"):
+                    set_nullable(key, key.lower() + "_file")
+                else:
+                    set_nullable(key, key.lower())
+
+    @classmethod
+    def _read_int_keywords(cls, content_dict, config_content):
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:133]: Is it possible to refactor this function such that it returns the dictionary?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 5]
```diff
@@ -1,263 +1,211 @@
...
+        schema = parser.add("ARG_TYPE")
+        schema.set_argc_minmax(2, 2)
+        schema.iset_type(0, ContentTypeEnum.CONFIG_INT)
+
+        return parser.parse(
+            config_file,
+        )
+
+    @classmethod
+    def _read_str_keywords(cls, content_dict, config_content):
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:118]: Is it possible to refactor this function such that it returns the dictionary?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 6]
```diff
@@ -1,263 +1,211 @@
...
+            raise ConfigValidationError(
+                f"Could not open job config file {config_file}"
+            ) from err
+
+        content_dict = {}
+
+        cls._read_str_keywords(content_dict, config_content)
+        cls._read_int_keywords(content_dict, config_content)
+
+        content_dict["executable"] = config_content.getValue("EXECUTABLE")
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:159]: I think this is already done as part of the `cls._read_str_keywords(content_dict, config_content)`

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 7]
```diff
@@ -1,263 +1,211 @@
...
+            name = os.path.basename(config_file)
+
+        try:
+            config_content = cls._parse_config_file(config_file)
+        except IOError as err:
+            raise ConfigValidationError(
+                f"Could not open job config file {config_file}"
+            ) from err
+
+        content_dict = {}
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:154]: It was a bit difficult to follow what this `content_dict` actually does I think I understand it but I also think that it would have been shorter and more clear if you just used the actually variables for the constructor like:

```python
    stdin_file: Optional[str] = config_content.get("STDIN")
    stdout_file: Optional[str] = config_content.get("STDOUT")
    stderr_file: Optional[str] = config_content.get("STDERR")
```
and implement the get inside the config_content.py something like: 

``` python
    def get(self, key):
        if not self.hasKey(key):
            return None
        value = self[key].getValue(-1, 0)
        if value == "null":
            return None
        return value
```
But I guess it is a matter of style

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 8]
```diff
@@ -1,263 +1,211 @@
...
+        schema.set_argc_minmax(2, 2)
+        schema.iset_type(0, ContentTypeEnum.CONFIG_INT)
+
+        return parser.parse(
+            config_file,
+        )
+
+    @classmethod
+    def _read_str_keywords(cls, content_dict, config_content):
+        def set_nullable(keyword, key):
```
Comment (DanSava) [inline src/ert/_c_wrappers/job_queue/ext_job.py:119]: This could be renamed to `might_set_value_none` also this function changes variables outside of its scope which makes it very hard to follow when reading the code

**Review by DanSava — CHANGES_REQUESTED** [review]

File: src/ert/_c_wrappers/enkf/res_config.py [hunk 9]
```diff
@@ -316,25 +317,24 @@ def _alloc_from_content(self, user_config_file=None, config=None):
         for job_description in config_content_dict.get(ConfigKeys.FORWARD_MODEL, []):
             job_name, args = parse_signature_job("".join(job_description))
             try:
-                job = self.installed_jobs[job_name].copy()
+                job = copy.deepcopy(self.installed_jobs[job_name])
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/res_config.py:320]: It calls the __deepcopy__() function.

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 10]
```diff
@@ -1,263 +1,211 @@
...
+        parser.add("EXEC_ENV").set_argc_minmax(1, 2)
+        parser.add("DEFAULT").set_argc_minmax(2, 2)
+        parser.add("ARGLIST").set_argc_minmax(1, -1)
+        schema = parser.add("ARG_TYPE")
+        schema.set_argc_minmax(2, 2)
+        schema.iset_type(0, ContentTypeEnum.CONFIG_INT)
+
+        return parser.parse(
+            config_file,
+        )
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/job_queue/ext_job.py:112]: this is so elegant and nice, LOVE IT

File: src/ert/_c_wrappers/enkf/res_config.py [hunk 11]
```diff
@@ -430,8 +430,8 @@ def _alloc_from_dict(self, config_dict):
                     f"Could not find job `{job_description[ConfigKeys.NAME]}`"
                     f" in list of installed jobs: {list(self.installed_jobs.keys())}"
                 ) from err
-            job.set_private_args_as_string(job_description.get(ConfigKeys.ARGLIST))
-            job.convertToCReference(None)
+            job.private_args = SubstitutionList()
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/res_config.py:433]: This was in order to replicate the existing `clear_private_args()` function. I might be able to simplify further, but perhaps in a different PR?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 12]
```diff
@@ -1,263 +1,211 @@
...
+            parser.add(path_key).set_argc_minmax(1, 1)
+
+        parser.add("EXECUTABLE", required=True).set_argc_minmax(1, 1)
+        parser.add("ENV").set_argc_minmax(1, 2)
+        parser.add("EXEC_ENV").set_argc_minmax(1, 2)
+        parser.add("DEFAULT").set_argc_minmax(2, 2)
+        parser.add("ARGLIST").set_argc_minmax(1, -1)
+        schema = parser.add("ARG_TYPE")
+        schema.set_argc_minmax(2, 2)
+        schema.iset_type(0, ContentTypeEnum.CONFIG_INT)
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/job_queue/ext_job.py:108]: one little remark - can we rename `schema` to `arg_type_schema`?

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 2 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:13]: ext_job fits that description perfectly though. There is just a static method that creates the object from a file.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 4 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:133]: Done

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 5 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:118]: Done

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 13]
```diff
@@ -1,263 +1,211 @@
...
+        def set_nullable(keyword, key):
+            value = config_content.getValue(keyword)
+            if value == "null":
+                value = None
+            content_dict[key] = value
+
+        for key in cls._str_keywords:
+            if config_content.hasKey(key):
+                if key in ("STDIN", "STDOUT", "STDERR"):
+                    set_nullable(key, key.lower() + "_file")
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/job_queue/ext_job.py]: this is a change in behaviour for `STDIN` (the C code didn't check for "null") - it's not present in current logs, and probably not much used, so i don't know how much we should care.

File: src/ert/_c_wrappers/enkf/res_config.py [hunk 14]
```diff
@@ -316,25 +317,24 @@ def _alloc_from_content(self, user_config_file=None, config=None):
...
         for job_description in config_content_dict.get(ConfigKeys.SIMULATION_JOB, []):
-            job = self.installed_jobs[job_description[0]].copy()
-            job.set_arglist(job_description[1:])
-            job.set_define_args(self.substitution_list)
-            job.clear_deprecated_argv()
-            job.convertToCReference(None)
+            job = copy.deepcopy(self.installed_jobs[job_description[0]])
+            job.arglist = job_description[1:]
+            job.define_args = self.substitution_list
+            # job.clear_deprecated_argv()
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/enkf/res_config.py:337]: No, fixed.

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 15]
```diff
@@ -1,263 +1,211 @@
...
+                else:
+                    set_nullable(key, key.lower())
+
+    @classmethod
+    def _read_int_keywords(cls, content_dict, config_content):
+        for key in cls._int_keywords:
+            if config_content.hasKey(key):
+                value = config_content.getValue(key)
+                if value > 0:
+                    # less than 0 in the config is equivalent to setting None
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/job_queue/ext_job.py:139]: `> 0` - `less than 0` - what about == 0?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 16]
```diff
@@ -1,263 +1,211 @@
...
+            # We unescape backslash here to keep backwards compatability ie. If
+            # the arglist contains a '\n' we interpret it as a newline.
+            content_dict["arglist"] = [
+                s.encode("utf-8", "backslashreplace").decode("unicode_escape")
+                for s in config_content["ARGLIST"][-1]
+            ]
+
+        arg_types_dict = defaultdict(lambda: ContentTypeEnum.CONFIG_STRING)
+        if "max_arg" in content_dict and content_dict["max_arg"] > 0:
+            arg_types_dict[content_dict["max_arg"] - 1] = ContentTypeEnum.CONFIG_STRING
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/job_queue/ext_job.py:172]: this one i don't understand (maybe because it's late now) - why is the type of the last argument set to config string?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 17]
```diff
@@ -1,263 +1,211 @@
...
+        for arg in config_content["ARG_TYPE"]:
+            arg_types_dict[arg[0]] = ContentTypeEnum(type_from_kw(arg[1]))
+        if arg_types_dict:
+            content_dict["arg_types"] = [
+                arg_types_dict[j]
+                for j in range(max(i for i in arg_types_dict.keys()) + 1)
```
Comment (valentin-krasontovitsch) [inline src/ert/_c_wrappers/job_queue/ext_job.py:178]: i have so many questions. again i'm in a rush to leave, and it's late, so maybe i'm doing the dumbs, but why not just use something like `len(arg_typed_dict)`?

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 18]
```diff
@@ -1,263 +1,211 @@
...
+            content_dict[key] = {}
+            if config_content.hasKey(keyword):
+                for env in config_content[keyword]:
+                    if len(env) > 1:
+                        content_dict[key][env[0]] = env[1]
+                    else:
+                        content_dict[key][env[0]] = None
+
+        set_env("environment", "ENV")
+        set_env("exec_env", "EXEC_ENV")
```
[the remark this answers is not in the document]
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py:193]: ah i see EXEC_ENV is covered in jobmanager. but ENV I didn't find checked directly, it's set in config file in `tests/unit_tests/c_wrappers/res/job_queue/test_forward_model_formatted_print.py`, but only used for a forward model and as far as i can see not checked on extjob or elsewhere. do you think you could add some check there?
in particular (am i being too pedantic now) to cover the unsetting of env vars mechanism?

File: tests/unit_tests/c_wrappers/res/job_queue/test_ext_job.py [hunk 19]
```diff
@@ -79,54 +81,32 @@ def test_load_forward_model_upgraded():
...
+        _ = ExtJob.from_config_file("CONFIG")
 
 
 @pytest.mark.usefixtures("use_tmpdir")
 def test_load_forward_model_execu_missing_raises():
     with open("CONFIG", "w") as f:
         f.write("EXECU missing_script.sh\n")
-    with pytest.raises(ValueError):
-        _ = ExtJob("CONFIG", True)
+    with pytest.raises(ConfigValidationError, match="CONFIG failed"):
```
Comment (valentin-krasontovitsch) [inline tests/unit_tests/c_wrappers/res/job_queue/test_ext_job.py:92]: can we match more on the error message? or is this just a generic failure?

**Review by valentin-krasontovitsch — CHANGES_REQUESTED** [review]
ah i got through it after all. very nice work! just a few questions / comments

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 2 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:13]: Also, we talked about eventually turning ResConfig into a pydantic model/dataclass. Then making ext_job into an equivalent of a dataclass is necessary.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 6 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:161]: Oups, it shouldnt as the keyword is special.

File: src/ert/_c_wrappers/job_queue/ext_job.py [hunk 20]
```diff
@@ -1,263 +1,211 @@
...
+        return replace(self)
+
+    _int_keywords = ["MAX_RUNNING", "MAX_RUNNING_MINUTES", "MIN_ARG", "MAX_ARG"]
+    _str_keywords = [
+        "STDIN",
+        "STDOUT",
+        "STDERR",
+        "START_FILE",
+        "EXECUTABLE",
+        "PORTABLE_EXE",
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:85]: It is. I removed it now.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 12 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:108]: Different PR perhaps?

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 17 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:176]: Yea, that is not the same :(

Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py:176]: could you explain a bit? do / can we not expect every argument (index) to be present in the dict?

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: thanks!

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 17 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:176]: The reason is that the file might not contain type information for all indecies.

File: tests/unit_tests/c_wrappers/res/enkf/test_res_config.py [hunk 21]
```diff
@@ -246,12 +246,8 @@ def test_extensive_config(setup_case):
 
         exp_job_data = snake_oil_structure_config["INSTALL_JOB"][job_name]
 
-        assert (
-            Path(exp_job_data["CONFIG"]).resolve()
-            == Path(job.get_config_file()).resolve()
-        )
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/enkf/test_res_config.py:252]: Exactly. Generally, not a good idea to keep the source of a data object as it complicates checking for equality and such.

File: tests/unit_tests/c_wrappers/res/job_queue/test_ext_job.py (hunk 19 above)
Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/job_queue/test_ext_job.py:92]: This is a generic failure for now. We will have to replace/improve the config parser to get better error messages.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:191]: Does `test_job_manager::test_exec_env` cover this case?

File: tests/unit_tests/c_wrappers/res/enkf/test_enkf_sim_model.py [hunk 22]
```diff
@@ -1,10 +1,18 @@
...
 import pytest
 
 from ert._c_wrappers.enkf import EnKFMain, ResConfig
 
 
+def valid_args(arg_types, arg_list: List[str], runtime: bool = False):
+    for arg, arg_type in zip(arg_list, arg_types):
+        if not arg_type.valid_string(arg, runtime):
+            return False
+    return True
```
[the remark this answers is not in the document]
Comment (eivindjahren) [reply tests/unit_tests/c_wrappers/res/enkf/test_enkf_sim_model.py]: Done

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 7 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:154]: I think that might be premature as these are the only field that interprets "null" as None.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 12 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: Ah, I understand now. Fixed.

Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: thank you : )

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 15 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: Fixed

File: tests/unit_tests/c_wrappers/res/enkf/test_res_config.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/enkf/test_res_config.py:252]: I see!

File: tests/unit_tests/c_wrappers/res/job_queue/test_ext_job.py (hunk 19 above)
Comment (valentin-krasontovitsch) [reply tests/unit_tests/c_wrappers/res/job_queue/test_ext_job.py:92]: alrighty, cool, thanks!

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 13 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: I guess its best to handle these consistently, even though this is potentially breaking change? The ones that would care would be those that set STDIN to null and expected there to be a file called null. Seems unrealistic. I have added it to the PR comment, title and mark, just in case. Will add it to the commit body aswell.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 16 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:170]: This is to ensure that arg_types list is at least as long as `max_arg`.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 13 above)
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: treating them consistently makes sense, for sure.
i also agree that it's unrealistic that this will hit anybody : ) the logs on azure do not contain STDIN at all...

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 16 above)
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py:170]: ah, and the last value is perhaps overwritten later, if supplied, right? that code is a bit difficult to read. maybe refactor at some later point... thanks for explaining!

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 3 above)
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py:91]: one solution that i could envision is to create a parser that lives in the class - not sure if that is the best solution though, and also whether it goes against Eivind's dataclass / as little logic as possible approach...

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 16 above)
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py]: I refactored it a bit so that it hopefully is more clear now.

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (eivindjahren) [reply src/ert/_c_wrappers/job_queue/ext_job.py:202]: Added

[the remark this answers is not in the document]
Comment (valentin-krasontovitsch) [reply src/ert/_c_wrappers/job_queue/ext_job.py:202]: cool thanks!

**Review by valentin-krasontovitsch — APPROVED** [review]

File: src/ert/_c_wrappers/job_queue/ext_job.py (hunk 3 above)
Comment (DanSava) [reply src/ert/_c_wrappers/job_queue/ext_job.py:91]: I guess this can be viewed as premature optimization. It is fine as it is
